src/index/IndexImpl.Text.cpp

```diff
@@ -116,26 +99,30 @@ void IndexImpl::logEntityNotFound(const string& word,
 }
 
 // _____________________________________________________________________________
-void IndexImpl::buildTextIndexFile(
-    const std::optional<std::pair<string, string>>& wordsAndDocsFile,
-    bool addWordsFromLiterals, TextScoringMetric textScoringMetric,
-    std::pair<float, float> bAndKForBM25) {
-  AD_CORRECTNESS_CHECK(wordsAndDocsFile.has_value() || addWordsFromLiterals);
+void IndexImpl::buildTextIndexFile(std::optional<const string> wordsFile,
+                                   std::optional<const string> docsFile,
+                                   bool addWordsFromLiterals,
+                                   bool useDocsFileForVocabulary,
+                                   bool addEntitiesFromWordsFile,
+                                   TextScoringMetric textScoringMetric,
+                                   std::pair<float, float> bAndKForBM25) {
+  bool addWordsFromFiles = docsFile.has_value() &&
+                           (wordsFile.has_value() || useDocsFileForVocabulary);
+  AD_CORRECTNESS_CHECK(addWordsFromFiles || addWordsFromLiterals);
+  AD_CORRECTNESS_CHECK(!addEntitiesFromWordsFile || useDocsFileForVocabulary);
+  string wordsFileString = wordsFile.value_or("");
+  string docsFileString = docsFile.value_or("");
```

Comment:
If you do this so early, then the ` wordsFileString` should be called ` wordsFile`  and the optional should be called something else.